// typescript-to-lua@1.31.1/dist/lualib-build/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isExportTableDeclaration = isExportTableDeclaration;
exports.isExportTable = isExportTable;
exports.isExportTableIndex = isExportTableIndex;
exports.isExportAlias = isExportAlias;
exports.isExportAssignment = isExportAssignment;
exports.isRequire = isRequire;
exports.isImport = isImport;
exports.isExportsReturn = isExportsReturn;
const tstl = require("..");
function isExportTableDeclaration(node) {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}
function isExportTable(node) {
    return tstl.isIdentifier(node) && node.text === "____exports";
}
function isExportTableIndex(node) {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}
function isExportAlias(node) {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}
function isExportAssignment(node) {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}
function isRequire(node) {
    return (tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require");
}
function isImport(node, importNames) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}
function isExportsReturn(node) {
    return (tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports");
}
